File: box-and-box/ports/python/resource.py
"""resource.py - Resource Arithmetic, faithful Python port (v0.8)

A Ledger is a CLOSED double-entry system - value is never created from nothing;
every spend is a transfer to a sink, every refill a transfer from a treasury - so
CONSERVATION holds structurally. Depletable resources obey linear logic (used
once); resources marked `reusable` (linear logic's `!` modality) can be used
freely without depletion. The novel bridge PRICES the repair operators. Laws
C1-C8, CB1-CB3. Mirrors resource.mjs.
"""

SINK = "#sink"
TREASURY = "#treasury"
FREE = "free"


class _Infeasible:
    __slots__ = ()

    def __repr__(self):
        return "INFEASIBLE"


INFEASIBLE = _Infeasible()


def _clone(L):
    return {
        "bal": {a: dict(r) for a, r in L["bal"].items()},
        "kind": dict(L["kind"]),
    }


def Ledger(bal=None, kind=None):
    return {"bal": bal if bal is not None else {}, "kind": kind if kind is not None else {}}


def balance(L, acct, res):
    return (L["bal"].get(acct) or {}).get(res, 0)


def total(L, res):
    return sum(r.get(res, 0) for r in L["bal"].values())


def transfer(L, res, frm, to, amt):
    """The one primitive: move amt of res between two accounts. Conserves the total."""
    if amt < 0 or balance(L, frm, res) < amt:
        return INFEASIBLE  # no overdraft (the affine floor)
    M = _clone(L)
    M["bal"].setdefault(frm, {})
    M["bal"].setdefault(to, {})
    M["bal"][frm][res] = M["bal"][frm].get(res, 0) - amt
    M["bal"][to][res] = M["bal"][to].get(res, 0) + amt
    return M


def spend(L, acct, res, amt):
    return transfer(L, res, acct, SINK, amt)  # consume -> sink


def refill(L, acct, res, amt):
    return transfer(L, res, TREASURY, acct, amt)  # accrue <- treasury


def affords(L, acct, cost):
    return all(balance(L, acct, res) >= amt for res, amt in cost.items())


def feasible(L, acct, cost):
    return affords(L, acct, cost)  # the alethic gate; else the action carries zero


def use(L, acct, res):
    """reusable (`!`) vs depletable: use depletes a depletable, never a reusable one."""
    if balance(L, acct, res) < 1:
        return {"ok": False, "L": L}
    if L["kind"].get(res) == "reusable":
        return {"ok": True, "L": L}  # copy freely - no depletion
    return {"ok": True, "L": spend(L, acct, res, 1)}  # linear - consumed exactly once


def allocate(L, task, amt):
    """continual learning: capacity is a CONSERVED resource (free -> committed)."""
    return transfer(L, "capacity", FREE, "task:" + task, amt)


def consolidate(L, task, mind="mind"):
    """mint reusable knowledge."""
    M = _clone(L)
    M["kind"]["know:" + task] = "reusable"
    M["bal"].setdefault(mind, {})
    M["bal"][mind]["know:" + task] = 1
    return M


def forget(L, task, mind="mind"):
    """reclaim capacity - only by releasing the knowledge."""
    amt = balance(L, "task:" + task, "capacity")
    M = transfer(L, "capacity", "task:" + task, FREE, amt)
    if M is INFEASIBLE:
        M = _clone(L)
    if M["bal"].get(mind) is not None:
        M["bal"][mind]["know:" + task] = 0  # no-free-reclaim tradeoff
    return M


def worthwhile(value, cost):
    return value >= cost


def repair(L, acct, resource="tokens", value=0, cost=0):
    """PRICING THE REPAIRS (Type II rationality): invoke only if affordable AND worth it."""
    if not affords(L, acct, {resource: cost}):
        return {"decision": "cannot-afford", "L": L}
    if not worthwhile(value, cost):
        return {"decision": "skip", "L": L}  # act on the current best
    return {"decision": "invoke", "L": spend(L, acct, resource, cost)}  # pay to deliberate / escalate

File: box-and-box/ports/go/resource.go
package main

// resource.go — Resource Arithmetic / linear-logic ledger (port of resource.mjs).
// Laws C1–C8, CB1–CB3. Closed double-entry system ⇒ conservation holds structurally.

const (
	SINK     = "#sink"
	TREASURY = "#treasury"
	FREE     = "free"
)

// Ledger : per-account balances + per-resource kind metadata.
type Ledger struct {
	Bal  map[string]map[string]float64
	Kind map[string]string
}

// NewLedger builds a Ledger.
func NewLedger(bal map[string]map[string]float64, kind map[string]string) Ledger {
	if bal == nil {
		bal = map[string]map[string]float64{}
	}
	if kind == nil {
		kind = map[string]string{}
	}
	return Ledger{Bal: bal, Kind: kind}
}

func cloneLedger(L Ledger) Ledger {
	bal := map[string]map[string]float64{}
	for a, r := range L.Bal {
		nr := map[string]float64{}
		for k, v := range r {
			nr[k] = v
		}
		bal[a] = nr
	}
	kind := map[string]string{}
	for k, v := range L.Kind {
		kind[k] = v
	}
	return Ledger{Bal: bal, Kind: kind}
}

func balance(L Ledger, acct, res string) float64 {
	if r, ok := L.Bal[acct]; ok {
		return r[res]
	}
	return 0
}

func total(L Ledger, res string) float64 {
	s := 0.0
	for _, r := range L.Bal {
		s += r[res]
	}
	return s
}

// transfer : move amt of res between two accounts. Conserves the grand total.
// Returns (ledger, ok). ok==false mirrors the JS INFEASIBLE sentinel.
func transfer(L Ledger, res, from, to string, amt float64) (Ledger, bool) {
	if amt < 0 || balance(L, from, res) < amt {
		return Ledger{}, false // INFEASIBLE — no overdraft
	}
	M := cloneLedger(L)
	if M.Bal[from] == nil {
		M.Bal[from] = map[string]float64{}
	}
	if M.Bal[to] == nil {
		M.Bal[to] = map[string]float64{}
	}
	M.Bal[from][res] = M.Bal[from][res] - amt
	M.Bal[to][res] = M.Bal[to][res] + amt
	return M, true
}

func spend(L Ledger, acct, res string, amt float64) (Ledger, bool) {
	return transfer(L, res, acct, SINK, amt)
}
func refill(L Ledger, acct, res string, amt float64) (Ledger, bool) {
	return transfer(L, res, TREASURY, acct, amt)
}

func affords(L Ledger, acct string, cost map[string]float64) bool {
	for res, amt := range cost {
		if balance(L, acct, res) < amt {
			return false
		}
	}
	return true
}

func feasible(L Ledger, acct string, cost map[string]float64) bool {
	return affords(L, acct, cost)
}

// UseResult is the result of use.
type UseResult struct {
	Ok bool
	L  Ledger
}

// use : depletes a depletable resource, but never a reusable one.
func use(L Ledger, acct, res string) UseResult {
	if balance(L, acct, res) < 1 {
		return UseResult{Ok: false, L: L}
	}
	if L.Kind[res] == "reusable" {
		return UseResult{Ok: true, L: L}
	}
	M, _ := spend(L, acct, res, 1)
	return UseResult{Ok: true, L: M}
}

// allocate : free → committed capacity. Returns (ledger, ok).
func allocate(L Ledger, task string, amt float64) (Ledger, bool) {
	return transfer(L, "capacity", FREE, "task:"+task, amt)
}

// consolidate : mint reusable knowledge.
func consolidate(L Ledger, task, mind string) Ledger {
	if mind == "" {
		mind = "mind"
	}
	M := cloneLedger(L)
	M.Kind["know:"+task] = "reusable"
	if M.Bal[mind] == nil {
		M.Bal[mind] = map[string]float64{}
	}
	M.Bal[mind]["know:"+task] = 1
	return M
}

// forget : reclaim capacity — only by releasing the knowledge.
func forget(L Ledger, task, mind string) Ledger {
	if mind == "" {
		mind = "mind"
	}
	amt := balance(L, "task:"+task, "capacity")
	M, ok := transfer(L, "capacity", "task:"+task, FREE, amt)
	if !ok {
		M = cloneLedger(L)
	}
	if M.Bal[mind] != nil {
		M.Bal[mind]["know:"+task] = 0
	}
	return M
}

func worthwhile(value, cost float64) bool { return value >= cost }

// RepairResult is the result of repair.
type RepairResult struct {
	Decision string // "cannot-afford" | "skip" | "invoke"
	L        Ledger
}

// repair : Type-II rationality — invoke a repair only if affordable AND worth it.
func repair(L Ledger, acct, resource string, value, cost float64) RepairResult {
	if resource == "" {
		resource = "tokens"
	}
	if !affords(L, acct, map[string]float64{resource: cost}) {
		return RepairResult{Decision: "cannot-afford", L: L}
	}
	if !worthwhile(value, cost) {
		return RepairResult{Decision: "skip", L: L}
	}
	M, _ := spend(L, acct, resource, cost)
	return RepairResult{Decision: "invoke", L: M}
}
